lib/backup/notifier/command.rb

```diff
@@ -0,0 +1,99 @@
+# encoding: utf-8
+
+module Backup
+  module Notifier
+    class Command < Base
+
+      ##
+      # Command to execute.
+      #
+      # Make sure it is accessible from your $PATH, or provide
+      # the absolute path to the command.
+      attr_accessor :command
+
+      ##
+      # Arguments to pass to the command.
+      #
+      # Must be an array of strings or callable objects.
+      #
+      # Callables will be invoked with #call(model, status),
+      # and the return value used as the argument.
+      #
+      # In strings you can use the following placeholders:
+      #
+      # %l - Model label
+      # %t - Model trigger
+      # %s - Status (success/failure/warning)
+      # %v - Status verb (succeeded/failed/succeeded with warnings)
+      #
+      # All placeholders can be used with uppercase letters to capitalize
+      # the value.
+      #
+      # Defaults to ["%L %v"]
+      attr_accessor :args
+
+      def initialize(model, &block)
+        super
+        instance_eval(&block) if block_given?
+
+        @args ||= ["%L %v"]
+      end
+
+      private
+
+      ##
+      # Notify the user of the backup operation results.
+      #
+      # `status` indicates one of the following:
+      #
+      # `:success`
+      # : The backup completed successfully.
+      # : Notification will be sent if `on_success` is `true`.
+      #
+      # `:warning`
+      # : The backup completed successfully, but warnings were logged.
+      # : Notification will be sent if `on_warning` or `on_success` is `true`.
+      #
+      # `:failure`
+      # : The backup operation failed.
+      # : Notification will be sent if `on_warning` or `on_success` is `true`.
+      #
+      def notify!(status)
+        IO.popen([@command] + args.map { |arg| format_arg(arg, status) })
```

Comment:
In other classes backup uses `Open4.popen4` https://github.com/backup/backup/blob/8c95dc77f8d7569a327a815fa561073e06c7e520/lib/backup/pipeline.rb#L54
Any particular reason you're using IO.open?

---

So but what exactly does the compiled command look like?

Like this?

```ruby
@command = "/bin/my-notifier"
@args = ["--model=%L", "--trigger=%T", "STATUS=%s"]
```

```sh
/bin/my-notifier --model=an-example --trigger=a-trigger STATUS=failure
```

(It would support ENV vars right?)

---

We've already implemented an ENV var in version 5 called `BACKUP_PROCEDURE` for some of the custom commands (hooks) you can run. (procedure == model) Some other data like path locations, like the db and logs, might also be send with ENV vars.

I chose this rather than building a custom command. It seems more portable that way. What do you think?